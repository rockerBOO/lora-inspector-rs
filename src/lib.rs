use std::fmt::Debug;
// use std::alloc::Global;
use std::string::String;

// use candle_core::safetensors;
// use std::error::Error;
// use std::marker::Send;
// use std::marker::Sync;

mod file;
mod metadata;
mod network;
mod norms;
mod weight;
mod worker;

pub type Result<T> = std::result::Result<T, InspectorError>;

#[derive(Debug)]
pub enum InspectorError {
    Candle(candle_core::Error),
    SafeTensor(safetensors::SafeTensorError),
    Load(String),
    Msg(String),
}

impl InspectorError {
    fn candle(err: candle_core::Error) -> InspectorError {
        InspectorError::Candle(err)
    }

    fn safetensor(err: safetensors::SafeTensorError) -> InspectorError {
        InspectorError::SafeTensor(err)
    }
}

impl From<candle_core::Error> for InspectorError {
    fn from(err: candle_core::Error) -> InspectorError {
        InspectorError::candle(err)
    }
}

impl From<safetensors::SafeTensorError> for InspectorError {
    fn from(err: safetensors::SafeTensorError) -> InspectorError {
        InspectorError::safetensor(err)
    }
}

pub fn get_base_name(name: &str) -> String {
    name.split('.')
        .filter(|part| !matches!(*part, "weight" | "lora_up" | "lora_down" | "alpha"))
        .fold(String::new(), |acc, v| {
            if acc.is_empty() {
                v.to_owned()
            } else {
                format!("{acc}.{v}")
            }
        })
}

#[cfg(test)]
mod tests {
    use crate::get_base_name;

    #[test]
    fn get_base_name_test() {
        let base_name = get_base_name("lora_unet_up_blocks_1_attentions_1_proj_out.lora_up.weight");
        assert_eq!(base_name, "lora_unet_up_blocks_1_attentions_1_proj_out");

        let base_name =
            get_base_name("lora_unet_up_blocks_1_attentions_1_proj_out.lora_down.weight");
        assert_eq!(base_name, "lora_unet_up_blocks_1_attentions_1_proj_out");

        let base_name = get_base_name("lora_unet_up_blocks_1_attentions_1_proj_out.alpha");
        assert_eq!(base_name, "lora_unet_up_blocks_1_attentions_1_proj_out");
    }
}
