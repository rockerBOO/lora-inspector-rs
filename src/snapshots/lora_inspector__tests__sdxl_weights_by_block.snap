---
source: src/lib.rs
expression: dbg!(result)
---
{
  "text_encoder": {
    "layer_27": 2.6981724264439193,
    "layer_02": 2.585883574878324,
    "layer_10": 2.5828950484584086,
    "layer_03": 2.5590740850556273,
    "layer_30": 2.8715523784561747,
    "layer_17": 2.856573554443203,
    "layer_08": 2.5526433805076016,
    "layer_04": 2.4886652477788647,
    "layer_11": 2.2360385661059925,
    "layer_24": 2.827270217244079,
    "layer_19": 2.918184613260589,
    "layer_21": 2.8049301858088875,
    "layer_05": 2.464273559417075,
    "layer_07": 2.5771845546871788,
    "layer_20": 2.7940594038519673,
    "layer_26": 2.90682780986022,
    "layer_25": 2.7996566374047074,
    "layer_09": 2.6051241374127776,
    "layer_15": 2.9749570401214847,
    "layer_23": 2.861240965268156,
    "layer_13": 2.8646155949651084,
    "layer_29": 2.798727564344976,
    "layer_16": 2.9225455991238527,
    "layer_22": 2.9790958653316757,
    "layer_14": 2.82534215634467,
    "layer_31": 3.8637997801833954,
    "layer_12": 2.909897691479945,
    "layer_18": 2.916504453428043,
    "layer_06": 2.5168113962228156,
    "layer_28": 2.906921097006807,
    "layer_01": 2.436261530287356,
    "layer_00": 2.398623786521228
  },
  "unet": {
    "output_01": 3.7485393705952412,
    "middle_00": 4.4912954465075785,
    "input_02": 3.787922562475381,
    "input_00": 3.936847869198427,
    "middle_01": 3.7100521284501666,
    "output_02": 4.57177559348496,
    "output_03": 5.532903021129882,
    "middle_02": 4.209200318584085,
    "input_01": 3.255837980451958,
    "middle_03": 7.846870343021596,
    "output_00": 4.158135129624979,
    "input_03": 4.395878374398331
  }
}
